### PR TITLE
[GenderBias] Add Controllable Gender Bias Task

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
Adds the `genderation_bias:controllable_task` task, which, given an input task, flattens and classifies each label as having male, female, or neutral words, and subsequently adds a control token to indicate such.

To use the teacher, simply run `parlai dd -t genderation_bias:controllable_task:<ORIGINAL_TASK>`, and the teacher takes care of the rest!

**Testing steps**
Tested with `parlai dd`; I additionally included a test to ensure that `flatten_and_classify`, the core of the teacher, works well.

**Logs**

```
$ pytest -k TestGenderationBiasTeacher
======= test session starts =====
platform linux -- Python 3.6.10, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /private/home/kshuster/ParlAI, configfile: pytest.ini, testpaths: tests
plugins: requests-mock-1.8.0
collected 408 items / 407 deselected / 1 selected

tests/tasks/test_genderation_bias.py .                                                                                                                                   [100%]

======= slowest 10 durations ======

(3 durations < 0.005s hidden.  Use -vv to show these durations.)
====== 1 passed, 407 deselected, 2 warnings in 5.05s ========
```

```
$ parlai dd -t genderation_bias:controllable_taks:convai2
...
- - - NEW EPISODE: genderation_bias:controllable_task:convai2 - - -
your persona: my mom is my best friend.
your persona: i have four sisters.
your persona: i believe that mermaids are real.
your persona: i love iced tea.
hi , how are you doing today ? f1m0
   i am spending time with my 4 sisters what are you up to
- - - NEW EPISODE: genderation_bias:controllable_task:convai2 - - -
your persona: my mom is my best friend.
your persona: i have four sisters.
your persona: i believe that mermaids are real.
your persona: i love iced tea.
hi , how are you doing today ?
i am spending time with my 4 sisters what are you up to
wow , four sisters . just watching game of thrones . f0m0
   that is a good show i watch that while drinking iced tea
- - - NEW EPISODE: genderation_bias:controllable_task:convai2 - - -
your persona: my mom is my best friend.
your persona: i have four sisters.
your persona: i believe that mermaids are real.
your persona: i love iced tea.
hi , how are you doing today ?
i am spending time with my 4 sisters what are you up to
wow , four sisters . just watching game of thrones .
that is a good show i watch that while drinking iced tea
i agree . what do you do for a living ? f0m0
   i'm a researcher i'm researching the fact that mermaids are real
17:28:32 | loaded 131438 episodes with a total of 131438 examples
```

### CODE REVIEW

File: parlai/tasks/genderation_bias/agents.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Generates a controllable_gen version of a ParlAI task, i.e., for tasks with multi-turn
+dialogues (episodes), this will generate a task with single example episodes, in which
+we append to the context a special classification token.
+
+In order to use this teacher, specify the task flag as follows:
+# `--task internal:gender_multiclass:controllable_task:<ORIGINAL TASK NAME>`.
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:12]: Remove this line?

File: parlai/tasks/genderation_bias/agents.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+class ControllableTaskTeacher(FixedDialogTeacher):
+    """
+    Generates a controllable_gen version of a ParlAI task, i.e., for tasks with multi-
+    turn dialogues (episodes), this will generate a task with single example episodes,
+    in which we append to the context a special classification token.
+    """
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        flattened = parser.add_argument_group('Flattening Args')
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:51]: Nit: maybe `ControllableTaskTeacher flattening arguments` or something?

File: parlai/tasks/genderation_bias/agents.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            help='How to join the dialogue history from previous turns.',
+        )
+        flattened.add_argument(
+            '--flatten-max-context-length',
+            type=int,
+            default=-1,
+            help='Maximum number of utterances to include per episode. '
+            'Default -1 keeps all.',
+        )
+        agent = parser.add_argument_group('Controllable Gen Teacher Args')
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:71]: Nit: `ControllableTaskTeacher arguments`?

File: parlai/tasks/genderation_bias/agents.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+            type='bool',
+            default=True,
+            help='Use 4 class classification for gendered word list (f0m0/f0m1/f1m0/f1m1). '
+            'Set false to use 3-class classification (MALE/FEMALE/NEUTRAL)',
+        )
+        agent.add_argument(
+            '--fixed-control',
+            type=str,
+            default='',
+            help='Always append this fixed control variable, good for deploy time.',
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:96]: hmm "variable" seems strange in this context (maybe because it's fixed, not variable, ha) - maybe "string"? Will leave this up to you

File: parlai/tasks/genderation_bias/agents.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        for task in tasks:
+            if hasattr(task, 'add_cmdline_args'):
+                task.add_cmdline_args(parser)
+
+        return parser
+
+    def __init__(self, opt: Opt, shared: TShared = None):
+        assert opt['flatten_delimiter'] == opt.get(
+            'delimiter', '\n'
+        ), '--flatten-delimiter and --delimiter are set differently, please inspect and set to the same to avoid unexpected results'
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:110]: 🎉

File: parlai/tasks/genderation_bias/agents.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        """
+        return os.path.join(
+            datapath,
+            f"{self.original_task_name.replace(':', '_')}_flattened_controllable_gen_{date}",
+        )
+
+    @classmethod
+    def build_wordlists(cls, opt: Opt) -> Tuple[List[str], List[str]]:
+        """
+        Load list of explicitly gendered words from.
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:149]: Hmm it looks like this sentence got split in two

File: parlai/tasks/genderation_bias/agents.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        :param filename:
+            name of (potentially) saved data
+
+        :return episodes:
+            return list of episodes, if available
+        """
+        # first check for the most recent date
+        save_dir = self._get_save_path(opt['datapath'], '*')
+        all_dates = []
+        for fle in glob.glob(os.path.join(save_dir, filename)):
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:253]: nit: `file`?

File: parlai/tasks/genderation_bias/agents.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+                f.write(json_data)
+            logging.info(f'[ Data successfully saved to path: {self.save_path} ]')
+        except Exception:
+            logging.warn('Data is not json serializable; not saving')
+
+    def get(self, episode_idx: int, entry_idx: int = 0) -> Message:
+        """
+        Return a flattened example.
+
+        If using a fixed control, put in instead of what was originally in the episode.
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/agents.py:301]: hmm maybe "put that in"?

File: parlai/tasks/genderation_bias/utils.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+) -> str:
+    """
+    Return a control token corresponding to gender within text.
+
+    :param text:
+        text to consider for control token
+    :param word_lists:
+        tuple of lists for male-specific and female-specific words
+    :param four_class:
+        whether to utilize four-class method (f0m0 --> f1m1) identifying.
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/utils.py:62]: Maybe rephrase this?

File: parlai/tasks/genderation_bias/utils.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+def flatten_and_classify(
+    episode: List[Message],
+    context_length: int,
+    word_lists: Tuple[List[str], List[str]],
+    include_labels: bool = True,
+    delimiter: str = '\n',
+    four_class: bool = False,
+):
+    """
+    Flatten the dialogue history of an episode, explore into N new examples.
```
Comment (EricMichaelSmith) [inline parlai/tasks/genderation_bias/utils.py:105]: hmm "explore" -> "divide"?

**Review by EricMichaelSmith — APPROVED** [review]
Looks good to me! Tiny nits

File: parlai/tasks/genderation_bias/README.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
+Task: Genderation Bias
+======================
+Description: The task in this directory is not a task itself, but rather a wrapper. The task will flatten a specified other ParlAI task (that is, turn multi-turn episodes into single-turn examples), and append a control token that corresponds with the level of gender present in the label. If `--four-class` is specified as `True`, the control token will be one of the following:
+
+- `f0m0` - no gender words in the label
```
Comment (emilydinan) [inline parlai/tasks/genderation_bias/README.md:5]: would note which word lists were used to generate these labels

File: parlai/tasks/genderation_bias/agents.py [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+            type=int,
+            default=-1,
+            help='If greater than zero, will stop building after a certain num of exs',
+        )
+        agent.add_argument(
+            '--four-class',
+            type='bool',
+            default=True,
+            help='Use 4 class classification for gendered word list (f0m0/f0m1/f1m0/f1m1). '
+            'Set false to use 3-class classification (MALE/FEMALE/NEUTRAL)',
```
Comment (emilydinan) [inline parlai/tasks/genderation_bias/agents.py:90]: we've never used m/f/n so maybe let's get rid of this option

**Review by emilydinan — COMMENTED** [review]
amazing, thanks so much for adding this 😻

File: parlai/tasks/genderation_bias/agents.py (hunk 12 above)
Comment (klshuster) [reply parlai/tasks/genderation_bias/agents.py:90]: will do - in that case @EricMichaelSmith make sure to update #3144 to remove the `--four-class` flag

File: parlai/tasks/genderation_bias/agents.py (hunk 7 above)
Comment (klshuster) [reply parlai/tasks/genderation_bias/agents.py:253]: `file` is a reserved name in Python but yeah `fle` is weird will change

File: parlai/tasks/genderation_bias/utils.py (hunk 10 above)
Comment (klshuster) [reply parlai/tasks/genderation_bias/utils.py:105]: oops meant `explode`

File: parlai/tasks/genderation_bias/agents.py (hunk 6 above)
Comment (klshuster) [reply parlai/tasks/genderation_bias/agents.py:149]: oh this might be black adding a period to the end of the top line of the docstring

**Review by emilydinan — APPROVED** [review]
